Rules/UseCompatibleTypes.cs

```diff
@@ -0,0 +1,854 @@
+            foreach (Ast definition in definitionAsts)
+            {
+                TypeDefinitionAst customType = (TypeDefinitionAst)definition;
+                customTypes.Add(customType.Name);
+            }
+
+            // These are namespaces used in the script.  This will help getting full type names for types
+            // from ast objects that only give us the type name as a string.  We will add these to the 
+            // beginning of our known namespaces list so we check those first.
+            IEnumerable<Ast> useStatementAsts = ast.FindAll(testAst => testAst is UsingStatementAst, true);
+            foreach (Ast use in useStatementAsts)
+            {
+                UsingStatementAst nspace = (UsingStatementAst)use;
+                knownNamespaces.Insert(0, nspace.Name.Value + ".");
+            }
+
+            // If we have no types to check, we can exit from this rule.
+            if (allTypesFromAst.Count.Equals(0))
+            {
+                return new DiagnosticRecord[0];
+            }
+
+            CheckCompatibility();
+
+            ProcessErrors();
+
+            return diagnosticRecords;
+        }
+
+        /// <summary>
+        /// Check if type is present in the target platform type list.
+        /// If not, create a Diagnostic Record for that type.
+        /// </summary>
+        private void CheckCompatibility()
+        {
+            foreach (dynamic typeObject in allTypesFromAst)
+            {
+                List<fullTypeNameObject> fullTypeNameObjectList = RetrieveFullTypeName(typeObject);
+
+                foreach (fullTypeNameObject nameObject in fullTypeNameObjectList)
+                {
+                    int couldNotResolveCount = 0;
+
+                    foreach (dynamic platform in psTypeMap)
+                    {
+                        if (nameObject.isCustomType)
+                        {
+                            // If this is a custom type, try and find it in our definition ast list.  If it's there
+                            // we can ignore it.
+                            if (customTypes.Contains(nameObject.fullName, StringComparer.OrdinalIgnoreCase))
+                            {
+                                continue;
+                            }
+                            // If this is not a known custom type, it could mean it is a type the user created with 
+                            // Add-Type or with a string variable containing the type information.  There is no reasonable
+                            // way to check for that, so create a diagnostic record.
+                            else
+                            {
+                                // If we have multiple target platforms to check and cannot resolve a type, we don't want
+                                // multiple 'could not resolve' errors for the same type.
+                                if (couldNotResolveCount < 1)
```

Comment:
If the only used values of `couldNotResolve` are 0 and (> 0), then a boolean might suit better here